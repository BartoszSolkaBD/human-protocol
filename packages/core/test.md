1. Can we use viaIR with foundry?

Contracts are compuling with viaIR using Foundry. 
You specify via-ir=true in Foundry.toml

2. If we start using viaIR, can we still upgrade deployed proxies?

Contracts were upgraded using viaIR in fact using Foundry. 


3. If we use viaIR, how will it affect to the gas cost?

Using viaIR, gas cost decreased slightly. 
Use : forge snapshot --diff with and without via-ir=true in Foundry.toml


4. If we move to Foundry, can we still upgrade deployed proxies?

Yes, deployed proxies were upgraded using Foundry. 