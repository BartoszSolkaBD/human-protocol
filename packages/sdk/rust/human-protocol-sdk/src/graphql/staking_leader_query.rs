use cynic::{impl_scalar};
use serde::{Serialize};


mod schema {
  cynic::use_schema!("schemas/leader.graphql");
}

impl_scalar!(i64, schema::I64);

#[derive(Serialize, Debug, Clone, cynic::QueryFragment)]
#[cynic(schema_path = "schemas/leader.graphql")]
pub struct Leader {
  id: String,
  address: String,
  amount_staked: cynic::Id,
  amount_allocated: cynic::Id,
  amount_locked: cynic::Id,
  locked_until_timestamp: cynic::Id,
  amount_withdrawn: cynic::Id,
  amount_slashed: cynic::Id,
  reputation: cynic::Id,
  reward: cynic::Id,
  amount_jobs_launched: cynic::Id,
  role: Option<String>,
  fee: Option<cynic::Id>,
  public_key: Option<String>,
  webhook_url: Option<String>,
  url: Option<String>,
}

#[derive(cynic::QueryVariables, Clone)]
pub struct LeaderFilter {
  pub id: String,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Root", variables = "LeaderFilter")]
#[cynic(schema_path = "schemas/leader.graphql")]
pub struct GetLeaderQuery {
  #[arguments(id: $id)]
  pub leader: Option<Leader>,
}

pub async fn run_query(url: &str, filter: &LeaderFilter) -> cynic::GraphQlResponse<GetLeaderQuery> {
  use cynic::http::ReqwestExt;

  let query = build_query(filter);
  
  reqwest::Client::new()
    .post(url)
    .run_graphql(query)
    .await
    .unwrap()
}

fn build_query(filter: &LeaderFilter) -> cynic::Operation<GetLeaderQuery, LeaderFilter> {
  use cynic::QueryBuilder;

  GetLeaderQuery::build(filter.clone())
}

#[cfg(test)]
mod test {
    use crate::{enums::ChainId, constants::NETWORKS};

    use super::*;
    use tokio::test;

    #[test]
    async fn snapshot_test_query() {
      let filter = LeaderFilter {
        id: "some_address".to_string()
      };
      let query = build_query(&filter);

      insta::assert_snapshot!(query.query);
    }

    #[tokio::test]
    async fn test_running_query() {
      let network = NETWORKS.get(&ChainId::PolygonMumbai).unwrap();
      let filter = LeaderFilter {
        id: "0x9d1f5d4c4aa1764bf63583c3392f626e2ffbe5f7".to_string()
      };
      let result = run_query(network.subgraph_url, &filter).await;
      if let Some(errors) = &result.errors {
          assert_eq!(errors.len(), 0);
      }
      insta::assert_debug_snapshot!(result.data);
    }

    #[tokio::test]
    async fn test_filter_builder() {
      let filter = LeaderFilter {
        id: "some_address".into()
      };

      assert_eq!(filter.id, "some_address".to_string());
    }
}