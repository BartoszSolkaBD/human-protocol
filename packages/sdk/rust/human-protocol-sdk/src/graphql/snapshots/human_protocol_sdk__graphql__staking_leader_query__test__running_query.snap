---
source: src/graphql/staking_leader_query.rs
expression: result.data
---
Some(
    GetLeaderQuery {
        leader: Some(
            Leader {
                id: "0x9d1f5d4c4aa1764bf63583c3392f626e2ffbe5f7",
                address: "0x9d1f5d4c4aa1764bf63583c3392f626e2ffbe5f7",
                amount_staked: Id(
                    "2",
                ),
                amount_allocated: Id(
                    "0",
                ),
                amount_locked: Id(
                    "0",
                ),
                locked_until_timestamp: Id(
                    "0",
                ),
                amount_withdrawn: Id(
                    "0",
                ),
                amount_slashed: Id(
                    "4",
                ),
                reputation: Id(
                    "0",
                ),
                reward: Id(
                    "0",
                ),
                amount_jobs_launched: Id(
                    "569",
                ),
                role: Some(
                    "Job Launcher",
                ),
                fee: None,
                public_key: None,
                webhook_url: None,
                url: None,
            },
        ),
    },
)
